### ISSUE TITLE
EnsureOakIndex: Does not create new indexes if OSGi configuration does not change

### ISSUE THREAD

**Issue by henrykuijpers** [Adobe-Consulting-Services/acs-aem-commons#1674]
## Required Information

* [x] AEM Version, including Service Packs, Cumulative Fix Packs, etc: AEM 6.2 SP1 CFP18
* [x] ACS AEM Commons Version: 3.6.2 (I know...)
* [x] Reproducible on Latest? yes

### Expected Behavior

When adding a new oak-index-definition (in /apps/project/oak-indexes), it should be picked up on deployment and executed (i.e. a new index created under /oak:index).

### Actual Behavior

New oak-index-definition is not picked up. Index is not created under /oak:index. The sling:osgiConfig remains the same, therefore the component is not re-activated/refreshed.

### Steps to Reproduce

1. Set up AEM 6.x
2. Install ACS AEM Commons newest
3. Create a sling:osgiConfig for EnsureOakIndex with config pointing to /apps/project/oak-indexes
4. Add an oak index definition under /apps/project/oak-indexes.
5. Notice that the definition is not picked up and also not installed.
6. Modify the sling:osgiConfig node.
7. Notice that the definition is now picked up and installed.

**Comment by henrykuijpers** [comment]
I believe that either of the following should be done to make this work:
1. Have a property "install=true" on the sling:osgiConfig node that is removed after installation - This allows to have your deployments trigger a (re)install of the indexes, consequently
2. A "watcher" for node events in /apps/project/oak-indexes, so changes are picked up and processed

@davidjgonzalez any thoughts on this?

**Comment by davidjgonzalez** [comment]
There is a JMX MBean that lets you invoke ensuring (unforced and forced) in a pinch.

I think the watcher is a bit more elegant (vs looking around and updating nodes under /apps, and handling the various osgi config formats.) We'd need to add a new property on `/apps/project/oak-indexes` to indicate it should be watched (`ensureOnChange` or something)...

we could probably auto-register it only for the paths defined in the main ensure osgi config, and only if they have the property set to true. There might be a corner-case w/ debouncing bulk changes (iirc if 3 nodes are updated, the watcher triggers thrice, once for each node changed - we'd only want to trigger a single reindex - which might naturally work itself out... just off the cuff right now :))

**Comment by joerghoh** [comment]
I am not a big fan of applying index definitions to their final location already during deployment.

* As soon as reindexing starts, you will have a (more or less) massive performance impact. That's something you don't want to encounter during a deployment, it makes the timing of the deployment unpredictable.
* When implementing it, you need to make sure that you don't run the ensureIndex in a JCR observation or sling event handler thread; synchronous index rebuilds might block this thread for hours.

That means, I don't want to make this a default behavior. Adding a special property to the index (e.g. ```ensureOnInstall``` or ```ènsureOnChange``` as suggested by @davidjgonzalez ) and starting the ensure process on deployment/change is ok for me, because it leaves it to you to negatively impact your timing.

**Comment by davidjgonzalez** [comment]
agree w everything @joerghoh said.

We could even put in a (configurable) delay (1 min default?) for when the ensuring kicks off to attempt to mitigate prematurely kicking off reindex while the app is restarting.

wrt to running in the observation hander thread i think we handle this already by using jobs to do the actual work, and @joerghoh smartly split light and heavy operations.

**Comment by joerghoh** [comment]
but that would still require a larger refactoring, because at the moment we handle everything at once; EnsureOakIndexJobHandler not designed to deal with individual definitions, but onyl with complete service instances.

**Comment by davidjgonzalez** [comment]
I think each service instance would just implement a listener for its `ensure-definitions.path`, and then "run" that service's "ensure" on change. ..  tbh, i think this flag could (should?) be added to the osgi config (vs the actual node). this is without looking at the code :) ... but might be relatively "easy" to add.

**Comment by joerghoh** [comment]
that would only trigger on change of an already existing EOI definition. But if I understand @henrykuijpers right he wants that a deployment automatically triggers a run of ensureOakIndex.

@henrykuijpers Have you tried to set the immediate=true property to the EnsureOakIndex instance? it will trigger the application immediately on startup of the component (thus immediately during the deployment). See https://github.com/Adobe-Consulting-Services/acs-aem-commons/blob/master/bundle/src/main/java/com/adobe/acs/commons/oak/impl/EnsureOakIndex.java#L127

**Comment by henrykuijpers** [comment]
@joerghoh, I saw this in the code, but I also saw that the default value of immediate = true already.

The problem is as follows: 
1. EnsureOakIndex definition (sling:osgiConfig) gets installed in the repo, EnsureOakIndex is executed for the path specified in the config 
2. Code deployments don't upgrade ACS Commons, nor restart its bundle, so the services (including EnsureOakIndex are not restarted, which is fine, nothing changed) 
3. Code deployments don't change the EnsureOakIndex sling:osgiConfig, so EnsureOakIndex is not triggering its @Modified or @Activate method (which is fine, nothing changed) 
4. Repository content that EnsureOakIndex config is pointing to, is updated, EnsureOakIndex is not triggered (which is not fine, I added a new index definition there on purpose).
